File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/severity/cross_role_leakage_guard.py
from __future__ import annotations



from .assertion_binding import (

    FLARE,

    CHRONIC,

    is_acute_chronic_cross,

    is_flare_or_chronic,

    resolve_assertion_binding,

)



CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON = "CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_BLOCKED"

TARGET_BINDING_MISSING_REASON = "TARGET_BINDING_MISSING"





def _cross_role_guard_options(input_contract: dict) -> dict:

    return {

        "label": input_contract.get("label"),

        "trusted_synthetic_binding_bypass": input_contract.get("trusted_synthetic_binding_bypass")

        is True,

    }





def _collapse_binding_failure(slot: dict, reason: str) -> dict:

    return {

        **slot,

        "severity_status": "NOT_EVALUATED",

        "severity_score": None,

        "severity_band": None,

        "severity_resolution_source": "NONE",

        "binding_status": "LEAKAGE_BLOCKED",

        "selected_cascade": None,

        "selected_dilution": None,

        "reason_codes": sorted(set([*slot["reason_codes"], reason])),

        "limitation_codes": sorted(set([*slot["limitation_codes"], "PHASE6_NO_NUMERIC_CASCADE"])),

    }





def apply_cross_role_leakage_guard(

    slots: list[dict], records: list[dict], input_contract: dict

) -> list[dict]:

    options = _cross_role_guard_options(input_contract)

    block_slot_ids: dict[str, str] = {}

    all_bindings: list[dict] = []



    for record in records:

        if record.get("patient_global_max_severity_label_only"):

            continue

        for assertion in record.get("severity_evidence_assertions") or []:

            binding = resolve_assertion_binding(assertion, record, options)

            all_bindings.append(binding)



            if binding["missing_acute_chronic_authority"]:

                block_slot_ids[binding["receiving_formula_slot_id"]] = TARGET_BINDING_MISSING_REASON

                continue



            if is_acute_chronic_cross(binding["bound_target_role"], binding["receiving_target_role"]):

                block_slot_ids[binding["receiving_formula_slot_id"]] = (

                    CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON

                )



            if binding["receiving_target_role"] == "STANDARD_FORMULA_TARGET" and is_flare_or_chronic(

                binding["bound_target_role"]

            ):

                block_slot_ids[binding["receiving_formula_slot_id"]] = (

                    CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON

                )



    by_evidence: dict[str, list[dict]] = {}

    for b in all_bindings:

        by_evidence.setdefault(b["evidence_item_id"], []).append(b)



    for occs in by_evidence.values():

        flare_recv = [o for o in occs if o["receiving_target_role"] == FLARE]

        chronic_recv = [o for o in occs if o["receiving_target_role"] == CHRONIC]

        if not flare_recv or not chronic_recv:

            continue



        if any(o["missing_acute_chronic_authority"] for o in occs):

            for o in occs:

                if is_flare_or_chronic(o["receiving_target_role"]):

                    block_slot_ids[o["receiving_formula_slot_id"]] = TARGET_BINDING_MISSING_REASON

            continue



        bound_roles = {o["bound_target_role"] for o in occs}

        each_authoritative = all(

            o["binding_authoritative"] and o["bound_target_role"] == o["receiving_target_role"]

            for o in occs

        )

        if each_authoritative and len(bound_roles) > 1:

            for o in occs:

                if is_flare_or_chronic(o["receiving_target_role"]):

                    block_slot_ids[o["receiving_formula_slot_id"]] = (

                        CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON

                    )



    if not block_slot_ids:

        return slots

    return [

        _collapse_binding_failure(slot, block_slot_ids[slot["formula_slot_id"]])

        if slot["formula_slot_id"] in block_slot_ids

        else slot

        for slot in slots

    ]

File: packages/clinical-contracts/src/rule4/severity/crossRoleLeakageGuard.ts
import type {
  Rule4FormulaSeverityRecord,
  Rule4SeverityAdapterInput,
  Rule4SlotSeverityResolution,
} from './types.js';
import {
  type CrossRoleGuardOptions,
  isAcuteChronicCross,
  isFlareOrChronic,
  resolveAssertionBinding,
  type ResolvedAssertionBinding,
} from './assertionBinding.js';

export const CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON =
  'CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_BLOCKED' as const;

export const TARGET_BINDING_MISSING_REASON = 'TARGET_BINDING_MISSING' as const;

const STANDARD = 'STANDARD_FORMULA_TARGET' as const;

function crossRoleGuardOptions(input: Rule4SeverityAdapterInput): CrossRoleGuardOptions {
  return {
    label: input.label,
    trustedSyntheticBindingBypass: input.trustedSyntheticBindingBypass === true,
  };
}

function collapseBindingFailure(
  slot: Rule4SlotSeverityResolution,
  reason: string,
): Rule4SlotSeverityResolution {
  return {
    ...slot,
    severityStatus: 'NOT_EVALUATED',
    severityScore: null,
    severityBand: null,
    severityResolutionSource: 'NONE',
    bindingStatus: 'LEAKAGE_BLOCKED',
    selectedCascade: null,
    selectedDilution: null,
    reasonCodes: [...new Set([...slot.reasonCodes, reason])].sort(),
    limitationCodes: [...new Set([...slot.limitationCodes, 'PHASE6_NO_NUMERIC_CASCADE'])].sort(),
  };
}

export { resolveAssertionBinding } from './assertionBinding.js';

/**
 * Acute/chronic cross-role leakage guard (Phase 6). Runs after per-slot resolve, before evidence binding gate.
 */
export function applyCrossRoleLeakageGuard(
  slots: Rule4SlotSeverityResolution[],
  records: readonly Rule4FormulaSeverityRecord[],
  input: Rule4SeverityAdapterInput,
): Rule4SlotSeverityResolution[] {
  const options = crossRoleGuardOptions(input);
  const blockSlotIds = new Map<string, string>();
  const allBindings: ResolvedAssertionBinding[] = [];

  for (const record of records) {
    if (record.patientGlobalMaxSeverityLabelOnly) {
      continue;
    }
    for (const assertion of record.severityEvidenceAssertions) {
      const binding = resolveAssertionBinding(assertion, record, options);
      allBindings.push(binding);

      if (binding.missingAcuteChronicAuthority) {
        blockSlotIds.set(binding.receivingFormulaSlotId, TARGET_BINDING_MISSING_REASON);
        continue;
      }

      if (isAcuteChronicCross(binding.boundTargetRole, binding.receivingTargetRole)) {
        blockSlotIds.set(
          binding.receivingFormulaSlotId,
          CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON,
        );
      }

      if (binding.receivingTargetRole === STANDARD && isFlareOrChronic(binding.boundTargetRole)) {
        blockSlotIds.set(
          binding.receivingFormulaSlotId,
          CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON,
        );
      }
    }
  }

  const byEvidence = new Map<string, ResolvedAssertionBinding[]>();
  for (const b of allBindings) {
    const list = byEvidence.get(b.evidenceItemId) ?? [];
    list.push(b);
    byEvidence.set(b.evidenceItemId, list);
  }

  for (const occs of byEvidence.values()) {
    const flareReceiving = occs.filter((o) => o.receivingTargetRole === 'CURRENT_ACUTE_FLARE');
    const chronicReceiving = occs.filter(
      (o) => o.receivingTargetRole === 'UNDERLYING_CHRONIC_TARGET',
    );
    if (flareReceiving.length === 0 || chronicReceiving.length === 0) {
      continue;
    }

    if (occs.some((o) => o.missingAcuteChronicAuthority)) {
      for (const o of occs) {
        if (isFlareOrChronic(o.receivingTargetRole)) {
          blockSlotIds.set(o.receivingFormulaSlotId, TARGET_BINDING_MISSING_REASON);
        }
      }
      continue;
    }

    const boundRoles = new Set(occs.map((o) => o.boundTargetRole));
    const eachAuthoritative = occs.every(
      (o) => o.bindingAuthoritative && o.boundTargetRole === o.receivingTargetRole,
    );
    if (eachAuthoritative && boundRoles.size > 1) {
      for (const o of occs) {
        if (isFlareOrChronic(o.receivingTargetRole)) {
          blockSlotIds.set(o.receivingFormulaSlotId, CROSS_ACUTE_CHRONIC_SEVERITY_LEAKAGE_REASON);
        }
      }
    }
  }

  if (blockSlotIds.size === 0) {
    return slots;
  }

  return slots.map((slot) => {
    const reason = blockSlotIds.get(slot.formulaSlotId);
    return reason ? collapseBindingFailure(slot, reason) : slot;
  });
}
